### PR TITLE
Initial implementation of state change callback

### PR DESCRIPTION
**Pull request by dlight**
This runs callbacks when the state machine attempts to change into `Download` and `Install`; callbacks may cancel the transition, returning the state machine to `Idle`.

### CODE REVIEW

**Comment by otavio** [comment]
Please add the callback to download, install, reboot.

Also, please add the tests so we ensure it works. I think it can be one generic test that handles the callback and an assert on each state that guarantees it will handle it.

What do you think?

**Comment by dlight** [comment]
Okay, sounds good.

File: src/states/mod.rs [hunk 1]
```diff
@@ -68,6 +70,34 @@ where
...
+{
+    fn handle_callbacks(self) -> Result<StateMachine> {
+        use states::transition::{state_change_callback, Transition};
+
+        // FIXME: Remove this clone
+        let firmware_path = &self.settings.firmware.metadata_path.clone();
+
+        let callback = self.callback_state_name();
+
+        let next_state = self.handle()?;
```
Comment (otavio) [inline src/states/mod.rs]: The `handle()` should only be run when the callback returns `Continue`. This is important as the state handling may be an expensive operation (e.g. `Install` state) so it should not run before we know if we can move to the next state.

Comment (otavio) [reply src/states/mod.rs]: I spent some time and ended up finding a possible way to solve this.

Please take a look at https://github.com/otavio/updatehub/commit/60736274adf0ea161f14243b45a9202f27a24f95

It still lacks tests but I liked the way it changes less code. Please let me know what do you think.

**Comment by otavio** [comment]
I am closing this PR as it has been superseded by #3

**Pull request closed without merging**